### ISSUE TITLE
Consider Switching to Py.test 

### ISSUE THREAD

**Issue by lkarthee** [mitsuhiko/jinja2#424]
Pytest is a Python test runner/library that lets you write powerful, concise tests with no boilerplate. It can run unittest, nose and doctest style tests and has over 100 plugins for easy integration with
many frameworks, editors and CI servers.

Boilerplate code like below can be eliminated from test case files from jinja2
```python 
# snippet from jinja2/testsuite/bytecode_cache.py
def suite():
suite = unittest.TestSuite()
suite.addTest(unittest.makeSuite(ByteCodeCacheTestCase))
return suite
```
Also the loaders like BetterLoader can be eliminated because py.test discovers test cases. Tests can be filtered by using options like `py.test -k expr ` to run tests matching expression.

```python
py.test test_mod.py   # run tests in module
py.test somepath      # run all tests below somepath
py.test -k stringexpr # only run tests with names that match the
                      # the "string expression", e.g. "MyClass and not method"
                      # will select TestMyClass.test_something
                      # but not TestMyClass.test_method_simple
py.test test_mod.py::test_func  # only run tests that match the "node ID",
                                # e.g "test_mod.py::test_func" will select
                                # only test_func in test_mod.py

```
http://pytest.org/latest/usage.html

Tests can be run parallely using xdist plugin. 
List of 3rd party plugins can be found at http://pytest.org/latest/plugins_index/index.html

Note: This issue is raised as part of 'adopt pytest month' initiative. http://pytest.org/latest/adopt.html

**Comment by auvipy** [comment]
+1

**Comment by untitaker** [comment]
Yeah, this is definetly planned. Pull requests welcome!

**Comment by auvipy** [comment]
for pr and migrating to pytest, prs should be send with converted tests written in pytest? what are the plans?

**Comment by untitaker** [comment]
Well, to be honest, there is no real plan. But I'd definetly merge a PR in the style of https://github.com/mitsuhiko/flask/pull/1165

**Comment by auvipy** [comment]
thats a very good reference to follow! would be very helpful! will try to send a PR in WIP mode then!

**Comment by untitaker** [comment]
Thanks!

Note that the checklist there is just a milestone-ish thing I was following... on the way I also converted multiple helper methods/functions to fixtures, whatever looked better.

**Comment by auvipy** [comment]
okey. jinja2 have less  tests then flask? I think so

**Comment by untitaker** [comment]
Not sure. Just start *somewhere*... if you get tired of it, I'm sure me or somebody else could continue.

**Comment by untitaker** [comment]
It seems that this has been suddenly tackled by @lkarthee: https://github.com/mitsuhiko/jinja2/pull/426

@lkarthee In the future, please leave a quick note if you're going to start on this, otherwise effort is duplicated.

**Comment by lkarthee** [comment]
@untitaker Sorry my mistake.

**Comment by untitaker** [comment]
Fixed by #426.
